src/include/storage/storage_structure/disk_array_collection.h

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <cstdint>
+
+#include "common/constants.h"
+#include "common/types/types.h"
+#include "storage/storage_structure/disk_array.h"
+
+namespace kuzu {
+namespace storage {
+
+class BufferManager;
+class WAL;
+class BMFileHandle;
+
+class DiskArrayCollection {
+    struct HeaderPage {
+        explicit HeaderPage(uint32_t numHeaders = 0)
+            : headers{}, nextHeaderPage{common::INVALID_PAGE_IDX}, numHeaders{numHeaders} {}
+        static constexpr size_t NUM_HEADERS_PER_PAGE =
+            (common::BufferPoolConstants::PAGE_4KB_SIZE - sizeof(common::page_idx_t) -
+                sizeof(uint32_t)) /
+            sizeof(DiskArrayHeader);
+
+        bool operator==(const HeaderPage&) const = default;
+
+        std::array<DiskArrayHeader, NUM_HEADERS_PER_PAGE> headers;
+        common::page_idx_t nextHeaderPage;
+        uint32_t numHeaders;
+    };
+    static_assert(std::has_unique_object_representations_v<HeaderPage>);
+
+public:
+    DiskArrayCollection(BMFileHandle& fileHandle, DBFileID dbFileID, BufferManager* bufferManager,
+        WAL* wal, common::page_idx_t firstHeaderPage = 0, bool bypassWAL = false);
+
+    static void writeEmptyHeadersToFile(FileHandle& handle, common::page_idx_t firstHeaderPage,
+        size_t numHeaders);
+
+    void prepareCommit();
+
+    void checkpointInMemory() {
+        for (size_t i = 0; i < headersForWriteTrx.size(); i++) {
```

Comment:
Should we avoid `size_t`? as I understand, it's not compatible on 32-bit platforms.

Reply:
It's not incompatible, it's just that we can't expect it to be 64-bit on a 32-bit platform, so it shouldn't be used for things like iterating over the elements in the disk array where we expect to be able to go beyond. But `std::vector::size` returns a `size_t` anyway (or at least usually does, according to cppreference, it's a `size_type` defined in the vector implementation).